### ISSUE TITLE
Rename `params` to `android.params`

### ISSUE THREAD

**Issue by grabbou** [rnpm/rnpm#141]
I believe params we currently ask should not be top-level but platform specific because of the following reasons:
- It's tricky to not ask for iOS or Android params when one of these projects is not in user app if they are shared. E.g. if you have an array of 3 questions, two related to your dependency installation on Android and one related to iOS (soon) - if end user has only iOS app there's no clear way to exclude that
- In most cases I believe they are going the answers user provides will be different (not shared) on iOS than on Android. In fact, it looks that currently react-native-code-push (cc: @lostintangent) already defines platform-specific message in the params question (`Enter your token for Android`) which signals they are going to define a separate question for iOS.

To avoid making breaking changes, we could deprecate top-level params and just assume they are Android params (`params = [...dependency.config.params, ...dependency.config.android.params];`) which we will remove later on.

**Comment by lostintangent** [comment]
This proposal makes a lot of sense to me!

**Comment by Kureev** [comment]
I actually think that we should keep `dependency.config.params` along with `ios.params` and `android.params`. 

> that was meant to be the same for both platforms (iOS & Android), he will now have to duplicate that and make the following:
```
rnpm: {
   { ios: params: [{title: "What is your token?"}] },
   { android: params: [{title: "What is your token?"}] },
}
```
but I believe better user experience is worth the additional few lines in package.json that will only have to be defined once.

Okay, true that if we're talking only about LOC. But we're talking about polling these parameters as well. If library assumes that user use the same token for iOS and Android (like API key or whatever), we shouldn't ask for it twice. So, here is my **addition** to your proposal:

We keep `dependency.config.params` for platform-shared params. If we have a case with a platform-specific params (like a GA tracking code), then you need to specify appropriate param in the platform config: `dependency.ios.params` or `dependency.android.params`.

In the case of parameter overlapping, platform specific params overrides shared params.

**Comment by grabbou** [comment]
That makes it impossible to solve the point #2 from my original issue. And I think asking twice is OK - in most cases you are going to have a different value. Even if we are about to ask twice - I am ok with that.

**Comment by Kureev** [comment]
I don't get how it makes your second point impossible:

> It's tricky to not ask for iOS or Android params when one of these projects is not in user app if they are shared. E.g. if you have an array of 3 questions, two related to your dependency installation on Android and one related to iOS (soon) - if end user has only iOS app there's no clear way to avoid asking for Android params.

The clear way to ignore linking process in general for missing part (i.e. if you have only Android, then iOS part would be completely skipped). The other thing is that you probably misread what I'm proposing: if you have 2 Android-specific params, you'll write them down in the `dependency.android.params` instead of having it in the `dependency.config.params`.

**Comment by grabbou** [comment]
Right, that is solvable, however I still don't like forcing people to define cross-platform value (because you can't predict whether that's going to be the case or not). If you have two questions (the same), one in Android, one in iOS polling for the same thing, we could suggest in the 2nd question the answer provided to the previous one (as a default value, hit enter to skip). That would work imo in the best way.

**Comment by Kureev** [comment]
> we could suggest in the 2nd question the answer provided to the previous one (as a default value, hit enter to skip). That would work imo in the best way.

I think that's a way more complicated solution than using 1 question for both platforms

**Comment by Kureev** [comment]
All in all, I think that should be up to the module author to decide that kind of things. I see benefits that it can bring to developers and don't see any issues that it may cause. 

So, what's your arguments against it?

**Comment by grabbou** [comment]
> So, what's your arguments against it, @grabbou?

Can you please double read what I said in last comment?

> I think that should be up to the module author to decide that kind of things (should it be two different questions or just a one for both platforms. And we shouldn't push our vision to users)

That's *exactly* pushing module author vision to the user. Whether it's shared or not should be always user responsibility.

**Comment by grabbou** [comment]
Another thing is that `rnpm` is mainly a tool for scaffolding so I would expect more users to either not use `params` (since by the time of installing they do not even have all the tokens in place) or just to answer once with a placeholder (or the same value) for the sake of speed (which in fact, is still easy to edit later).

Keeping that in mind, I can potentially agree on leaving the shared, android / ios & the overriding proposal.

**Comment by Kureev** [comment]
Yeah! And I'm always open to revisit this solution based on the user's feedback. I can be wrong, but for now it seems to be a smallest move that will bring us lots of value. 

**So just to summarize what we discussed:**
- We introduce a platform-specific parameters for the accurate config;
- We introduce a common parameters that are intended to scaffold library correctly and/or setup a shared value across two platforms
- Platform-specific parameters > common parameters which means that if you have a common parameter called "A" and platform-specific parameter called the same("A"), then during the linking process platform-specific parameter will override a common one.

**Comment by grabbou** [comment]
That's cool. Will require few updates in how we check for `isInstalled` before asking, but it's doable. Will submit a patch after we get rewritten android merged.

**Comment by Kureev** [comment]
Can we determine a next actionable step, @grabbou ? I can pick it up

**Comment by grabbou** [comment]
I am afraid we can't, since you disagree with me and it's a tie. I'd like someone else to chime in and provide more context if possible.

**Comment by Kureev** [comment]
I think both approaches that we described does make sense. All in all, implementation details shouldn't delay a feature. So, I'm fine with asking twice for the same parameter for different platform.

That said, we can move forward with your approach:

- Move `params` to `android.params`
- Create `ios.params`
- Every time when we ask for param use prefix in the message. Example:
  
  - `[iOS] Input your GA token:`
  - `[Android] Input your GA token:`

Would it be a solution that fits you?

**Comment by grabbou** [comment]
Yeah, I think that would work :) Do you think we should PR against this repo or against the react-native one?

**Comment by Kureev** [comment]
I think we should merge https://github.com/facebook/react-native/pull/7899 in a first place and then freeze this repo.
